### PR TITLE
implement wal interface

### PR DESCRIPTION
**Pull request by MarinPostma**
implement the sqlite Wal interface using the `SharedWal`

### CODE REVIEW

File: libsql-wal/src/wal.rs [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+
+        tracing::debug!(invalidate_cache, "read started");
+        Ok(invalidate_cache)
+    }
+
+    #[tracing::instrument(skip_all, fields(id = self.conn_id))]
+    fn end_read_txn(&mut self) {
+        match self.tx.take() {
+            Some(Transaction::Read(tx)) => tx,
+            Some(Transaction::Write(tx)) => tx.downgrade(),
```
Comment (avinassh) [inline libsql-wal/src/wal.rs:115]: in which situation a write transaction can call `end_read_txn`?

File: libsql-wal/src/wal.rs [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+    #[tracing::instrument(skip_all, fields(id = self.conn_id))]
+    fn undo<U: libsql_sys::wal::UndoHandler>(
+        &mut self,
+        handler: Option<&mut U>,
+    ) -> libsql_sys::wal::Result<()> {
+        match self.tx {
+            Some(Transaction::Write(ref mut tx)) => {
+                if tx.is_commited() {
+                    return Ok(());
+                }
```
Comment (avinassh) [inline libsql-wal/src/wal.rs:196]: shouldn't this return an error since you are calling undo on a committed transaction

**Review by avinassh — APPROVED** [review]

File: libsql-wal/src/wal.rs (hunk 1 above)
Comment (MarinPostma) [reply libsql-wal/src/wal.rs:115]: Usually, SQLite will end the write and then end the read, but in OG SQLite WAL implementation, ending a read transaction also ends the write

File: libsql-wal/src/wal.rs (hunk 2 above)
Comment (MarinPostma) [reply libsql-wal/src/wal.rs:196]: That's what I initially thought, but SQLite can call undo after a commit, and that's noop. I found that with fault injection.

File: libsql-wal/src/wal.rs (hunk 1 above)
Comment (MarinPostma) [reply libsql-wal/src/wal.rs:115]: Although ending a write txn does not end a read txn

**Pull request merged**